<a href="https://colab.research.google.com/github/heisenberg304/Housing-Prices-prediction/blob/develop/Modeling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Modeling

│

├── Dummy Regressor


├── Linear Regression


├── Ridge Regression


├── Lasso Regression


├── Decision Tree Regressor


├── Random Forest Regressor


├── Gradient Boosting


└── XGBoost

#LOAD DATA

In [1]:
%%writefile data_loading.py
import pandas as pd
from sklearn.model_selection import train_test_split

#cargar datos
# data train no filled
train_not_filled = pd.read_csv("/content/drive/MyDrive/house_prices_advanced/HP_train_not_filled_encoding.csv")

# data train filled
train_filled = pd.read_csv("/content/drive/MyDrive/house_prices_advanced/HP_train_filled_encoding.csv")

#x,y
y = train_filled["SalePrice"]
x = train_filled.drop("SalePrice", axis=1)

#x,y
y_nf = train_not_filled["SalePrice"]
x_nf = train_not_filled.drop("SalePrice", axis=1)


def split_dataset(x, y):  #split 80/20
  x_train, x_test, y_train, y_test = train_test_split(
      x,
      y,
      test_size = 0.2,
      random_state=42
  )
  return x_train, x_test, y_train, y_test
x_train, x_test, y_train, y_test = split_dataset(x, y)
x_train_nf, x_test_nf, y_train_nf, y_test_nf = split_dataset(x_nf, y_nf)

Writing data_loading.py


In [23]:
%%writefile modeling_def.py
from sklearn.metrics import mean_squared_error, root_mean_squared_error, mean_absolute_error, r2_score
from sklearn.model_selection import KFold, cross_val_score

#entrenar modelo
def entrenar_model(model, x, y):
  return model.fit(x,y)

#predicciones del modelo
def predict_model(model, x):
  return model.predict(x)

#metricas del modelo
def evaluate_model(y_true, y_pred):
  MSE = mean_squared_error(y_true, y_pred)
  RMSE = root_mean_squared_error(y_true, y_pred)
  MAE = mean_absolute_error(y_true, y_pred)
  R2 = r2_score(y_true, y_pred)
  return MSE, RMSE, MAE, R2

#creando kflod
def k_fold():
  kfold = KFold(n_splits=5,
                shuffle=True,
                random_state=42)
  return kfold

#validacion cruzada
def cross_val(x_train, y_train, model, kfold):
  cv = cross_val_score(
        estimator=model,
        X=x_train,
        y=y_train,
        cv=kfold,
        scoring="neg_root_mean_squared_error"
        )
  return cv


def cv_mean_std(cv):
  cv = (cv * -1)
  cv_std = round(float(cv.std()), 3)
  cv_mean = round(float(cv.mean()), 3)
  return cv_std, cv_mean


Writing modeling_def.py


#Baseline
   - Dummy

In [ ]:
#Dummy Regressor
from data_loading import x_train, y_train, x_test, y_test
from sklearn.dummy import DummyRegressor
from modeling_def import entrenar_model, predict_model, evaluate_model

#modelo
model = DummyRegressor()

#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(y_test, predict)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')


#Linear Models
   - Linear
   - Ridge
   - Lasso

In [ ]:
#Linear Regression
from sklearn.linear_model import LinearRegression
from data_loading import x_train, y_train, x_test, y_test
from modeling_def import entrenar_model, predict_model, evaluate_model

#modelo
model = LinearRegression()

#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)


#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(y_test, predict)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')


In [27]:
#Ridge Regression
from sklearn.linear_model import Ridge
from data_loading import x_train, y_train, x_test, y_test
from modeling_def import entrenar_model, predict_model, evaluate_model, k_fold, cross_val, cv_mean_std

#modelo
model = Ridge(alpha=10)

#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(y_test, predict)

#validacion cruzada
kfold = k_fold()
cv = cross_val(x_train, y_train, model, kfold)
cv_std, cv_mean = cv_mean_std(cv)

#print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \n Mean Squared Error: {MSE}')

print(f'Promedio validacion cruzada: {cv_mean} \nDesviacion estandar {cv_std}')

Promedio validacion cruzada: 32020.029 
Desviacion estandar 7546.345


In [28]:
#Lasso Regression
from sklearn.model_selection import KFold, cross_val_score

from sklearn.linear_model import Lasso
from data_loading import x_train, y_train, x_test, y_test
from modeling_def import entrenar_model, predict_model, evaluate_model, k_fold, cross_val, cv_mean_std

#modelo
model = Lasso(alpha=200)

#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(y_test, predict)

#validacion cruzada
kfold = k_fold()
cv = cross_val(x_train, y_train, model, kfold)
cv_std, cv_mean = cv_mean_std(cv)

#print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \n Mean Squared Error: {MSE}')

print(f'Promedio validacion cruzada: {cv_mean} \nDesviacion estandar {cv_std}')

Promedio validacion cruzada: 32073.645 
Desviacion estandar 7937.017


#Tree Models
   - Decision Tree Regressor
   - Random Forest Regressor

In [29]:
#Decision Tree Regressor
from sklearn.tree import DecisionTreeRegressor
from data_loading import x_train, y_train, x_test, y_test
from modeling_def import entrenar_model, predict_model, evaluate_model, k_fold, cross_val, cv_mean_std

#modelo
model = DecisionTreeRegressor(max_depth=7,          #7
                              min_samples_split=3,  #3
                              min_samples_leaf=1,   #1
                              random_state=7
)

#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(y_test, predict)

#validacion cruzada
kfold = k_fold()
cv = cross_val(x_train, y_train, model, kfold)
cv_std, cv_mean = cv_mean_std(cv)

#print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \n Mean Squared Error: {MSE}')

print(f'Promedio validacion cruzada: {cv_mean} \nDesviacion estandar {cv_std}')

Promedio validacion cruzada: 38768.336 
Desviacion estandar 4323.641


In [30]:
#Random Forest Regressor
from sklearn.ensemble import RandomForestRegressor
from data_loading import x_train, y_train, x_test, y_test
from modeling_def import entrenar_model, predict_model, evaluate_model, k_fold, cross_val, cv_mean_std

#modelo
model = RandomForestRegressor(n_estimators= 150, #100-150
                              #max_depth= ,      #default
                              min_samples_leaf=1,#default
                              min_samples_split=2, #default
                              max_features=None, #default
                              random_state= 3
)
#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(y_test, predict)

#validacion cruzada
kfold = k_fold()
cv = cross_val(x_train, y_train, model, kfold)
cv_std, cv_mean = cv_mean_std(cv)

#print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \n Mean Squared Error: {MSE}')

print(f'Promedio validacion cruzada: {cv_mean} \nDesviacion estandar {cv_std}')

Promedio validacion cruzada: 30665.098 
Desviacion estandar 4980.019


#Boosting Models
   - Gradient Boosting Regressor
   - XGBoost Regressor

In [31]:
#Gradient Boosting Regressor
from sklearn.ensemble import GradientBoostingRegressor
from data_loading import x_train, y_train, x_test, y_test
from modeling_def import entrenar_model, predict_model, evaluate_model, k_fold, cross_val, cv_mean_std

#modelo
model = GradientBoostingRegressor(learning_rate=0.08, #0.07 - 0.08
                                  n_estimators=200, #200
                                  max_depth=6, #6
                                  subsample=0.9, #0.8 - 0.9
                                  random_state=5
)
#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(y_test, predict)

#validacion cruzada
kfold = k_fold()
cv = cross_val(x_train, y_train, model, kfold)
cv_std, cv_mean = cv_mean_std(cv)

#print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \n Mean Squared Error: {MSE}')

print(f'Promedio validacion cruzada: {cv_mean} \nDesviacion estandar {cv_std}')

Promedio validacion cruzada: 28870.457 
Desviacion estandar 5383.561


In [32]:
#XG Boost Regressor
from xgboost import XGBRegressor
from data_loading import x_train, y_train, x_test, y_test
from modeling_def import entrenar_model, predict_model, evaluate_model, k_fold, cross_val, cv_mean_std

#modelo
model = XGBRegressor(learning_rate=0.11, #0.1
                     n_estimators=170, #150-180
                     max_depth=6, #6
                     subsample=0.9, #0.9
                     colsample_bytree=0.8, #0.8
                     gamma=1, #No hace nada
                     min_child_weight=2, #2
                     random_state= 5
)

#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(y_test, predict)

#validacion cruzada
kfold = k_fold()
cv = cross_val(x_train, y_train, model, kfold)
cv_std, cv_mean = cv_mean_std(cv)

#print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \n Mean Squared Error: {MSE}')

print(f'Promedio validacion cruzada: {cv_mean} \nDesviacion estandar {cv_std}')

Promedio validacion cruzada: 28429.844 
Desviacion estandar 4711.865


In [33]:
#XG Boost Regressor
from xgboost import XGBRegressor
from data_loading import y_train_nf, x_train_nf, y_test_nf, x_test_nf
from modeling_def import entrenar_model, predict_model, evaluate_model, k_fold, cross_val, cv_mean_std

#modelo
model = XGBRegressor(learning_rate=0.15, #0.1
                     n_estimators=190, #150-200
                     max_depth=6, #6
                     subsample=0.9, #0.9
                     colsample_bytree=1, #default
                     gamma=1, #No hace nada
                     min_child_weight=2, #2
                     random_state= 5
)
#entrenar modelo
modelo_entrenado_nf = entrenar_model(model, x_train_nf, y_train_nf)

#predicciones del modelo
predict_nf = predict_model(modelo_entrenado_nf, x_test_nf)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(y_test_nf, predict_nf)

#validacion cruzada
kfold = k_fold()
cv = cross_val(x_train_nf, y_train_nf, model, kfold)
cv_std, cv_mean = cv_mean_std(cv)

#print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \n Mean Squared Error: {MSE}')

print(f'Promedio validacion cruzada: {cv_mean} \nDesviacion estandar {cv_std}')

Promedio validacion cruzada: 29148.083 
Desviacion estandar 5746.988


#Tabla comparativa: Baseline & Linear Models
---

| Linear Model | RMSE ↓ | MAE ↓ | R² ↑ |
|---------|--------:|-------:|------:|
| Dummy | 87619.03 | 62575.926 | -0.00088 |  |  |
| Linear Regression| 83119.01 | 23911.31 | 0.099286 |
| Ridge (alpha=0.1) | 32616.5 | 19495.6 | 0.86130 |
| Ridge (default) | 30502.54 | 19735.23 | 0.87870 |
| Ridge (alpha=10)| 30556.64 | 18873.65 | 0.87827 |
| Ridge (alpha=100) | 32563.56 | 18873.41 | 0.86175 |
| Lasso (alpha=0.1) | 36125.65 | 19674.17 | 0.8299 |
| Lasso (default) | 35364.56 | 19381.04 | 0.83695 |
| Lasso (alpha=10) | 34986.26 | 18713.33 | 0.84042 |
| Lasso (alpha=100) | 31226.81 | 18633.83 | 0.87287 |
| Lasso (alpha=200) | 31110.26 | 18271.8 | 0.87382 |
---



#Tabla comparativa: Tree Models
---

| Tree Model | RMSE ↓ | MAE ↓ | R² ↑ |
|---------|--------:|-------:|------:|
| Decision Tree (default) | 40969.49 | 27266.78 | 0.78117 |
| Decision Tree (best) | 35597.6 | 24211.36 | 0.83479 |
| Random Forest (default) | 29235.14 | 17714.42 | 0.88857 |
| Random Forest (best) | 28906.58 | 17504.08 | 0.89106 |

---



#Tabla comparativa: Boosting Models
---

| Tree Model | RMSE ↓ | MAE ↓ | R² ↑ |
|---------|--------:|-------:|------:|
| Gradient Boosting (default) | 28656.91 | 17391.25 | 0.89294 |
| Gradient Boosting (best) | 25698.39 | 16015.58 | 0.91390 |
| XG Boost (default) | 27970.32 | 18042.61 | 0.89800 |
| XG Boost (best) | 24100.17 | 15489.14 | 0.92428 |
| XGBoost NaN handled natively (default) | 28267.21 | 17823.19 | 0.89583 |
| XGBoost NaN handled natively (best) | 24567.6 | 15939.87 | 0.92131 |

---

#Tabla comparativa de los mejores modelos
---

| Modelo | RMSE ↓ | MAE ↓ | R² ↑ | CV RMSE | CV STD |
|---------|--------:|-------:|------:|------:|------:|
| Ridge Regression| 30556.638 | 18873.65 | 0.87827 | 32020.029  | 7546.345 |
| Lasso Regression| 31110.26 | 18271.797 | 0.8738 | 32073.645 | 7937.017 |
| Decision Tree Regressor| 35597.6 | 24211.356 | 0.83479 | 38768.336 | 4323.641 |
| Random Forest Regressor| 28906.5 | 17504.08 | 0.891062 | 30665.098 | 4980.019 |
| Gradient Boosting Regressor| 25698.39 | 16015.58 | 0.91390 | 28870.457 | 5383.561 |
| **XGBoost Regressor**| **24100.17** | **15489.14** | **0.92428** | **28429.844** | **4711.865** |
| XGBoost NaN handled natively| 24567.6 | 15939.87 | 0.92131 | 29148.083 | 5746.988 |
---
